### PR TITLE
Fix for guti based registration and T3512

### PR DESCRIPTION
**Pull request by VinayKumar1591**
[5G: MVC][agw] Fix for guti based registration and sending T3512 in Registration accept

## Summary

Added conditions to support GUTI based registration

## Test Plan

Tested with NGAP Tester and verified E2E call flow

### CODE REVIEW

File: lte/gateway/c/oai/tasks/amf/amf_recv.cpp [hunk 1]
```diff
@@ -118,6 +118,7 @@ int amf_handle_registration_request(
     OAILOG_DEBUG(LOG_NAS_AMF, "Processing REGITRATION_REQUEST message\n");
   }
   amf_registration_request_ies_t* params = new (amf_registration_request_ies_t);
+  memset(params, 0, sizeof(amf_registration_request_ies_t));
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_recv.cpp]: This declaration should zero initialize: `amf_registration_request_ies_t* params = new amf_registration_request_ies_t();`

File: lte/gateway/c/oai/tasks/amf/amf_recv.cpp [hunk 2]
```diff
@@ -118,6 +118,7 @@ int amf_handle_registration_request(
     OAILOG_DEBUG(LOG_NAS_AMF, "Processing REGITRATION_REQUEST message\n");
   }
   amf_registration_request_ies_t* params = new (amf_registration_request_ies_t);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_recv.cpp]: when is this dynamically allocated memory gets freed up?

Comment (VinayKumar1591) [reply lte/gateway/c/oai/tasks/amf/amf_recv.cpp:120]: This should be freed when UE moved to DEREGISTERED

Comment (VinayKumar1591) [reply lte/gateway/c/oai/tasks/amf/amf_recv.cpp]: once procedure is completed and working on to resolve this

File: lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp [hunk 3]
```diff
@@ -432,6 +432,18 @@ int amf_proc_security_mode_control(
...
+      // int r = sprintf(
+      //   snni, "5G:mnc%01d%01d%01d.mcc%01d%01d%01d.3gppnetwork.org",
+      //  (char*) plmn.mnc_digit1, (char*) plmn.mnc_digit2,
+      // (char*) plmn.mnc_digit3, (char*) plmn.mcc_digit1,
+      //  (char*) plmn.mcc_digit2, (char*) plmn.mcc_digit3);
+      //  if (r != 32) {
+      //    OAILOG_ERROR(LOG_NAS_AMF, "failed to construct SNN!\n");
+      //  OAILOG_FUNC_RETURN(LOG_NAS_AMF, RETURNerror);
+      // } else {
+      OAILOG_ERROR(LOG_NAS_AMF, " SNN %s\n", snni);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp]: why is this ERROR level?

File: lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp [hunk 4]
```diff
@@ -432,6 +432,18 @@ int amf_proc_security_mode_control(
...
 
+      // int r = sprintf(
+      //   snni, "5G:mnc%01d%01d%01d.mcc%01d%01d%01d.3gppnetwork.org",
+      //  (char*) plmn.mnc_digit1, (char*) plmn.mnc_digit2,
+      // (char*) plmn.mnc_digit3, (char*) plmn.mcc_digit1,
+      //  (char*) plmn.mcc_digit2, (char*) plmn.mcc_digit3);
+      //  if (r != 32) {
+      //    OAILOG_ERROR(LOG_NAS_AMF, "failed to construct SNN!\n");
+      //  OAILOG_FUNC_RETURN(LOG_NAS_AMF, RETURNerror);
+      // } else {
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp]: remove the unused code.

File: lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp [hunk 5]
```diff
@@ -331,7 +331,7 @@ int amf_proc_security_mode_control(
   int amf_eea = 0;
   int amf_eia = 0;  // Integrity Algorithm 2
   amf_plmn_t plmn;
-  uint8_t snni[32]  = {0};
+  uint8_t snni[40]  = {0};
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp]: no need to change to 40 as this is 32 bytes.

File: lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp (hunk 4 above)
Comment (ulaskozat) [reply lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp]: why do you have (char *) casting in the commented code..

File: lte/gateway/c/oai/tasks/amf/amf_as.cpp [hunk 6]
```diff
@@ -1135,10 +1135,11 @@ static int amf_as_security_req(
...
               &ue_context->amf_context._security;
-          amf_security_context->selected_algorithms.integrity =
-              0;  // TODO get this computed
-          amf_security_context->selected_algorithms.encryption =
-              0;  // TODO get this computed
+          amf_security_context->selected_algorithms.integrity = 0;
+          //   M5G_NAS_SECURITY_ALGORITHMS_128_5G_IA2;  // TODO get this
+          //   computed
+          amf_security_context->selected_algorithms.encryption = 0;
+          //  M5G_NAS_SECURITY_ALGORITHMS_5G_EA0;  // TODO get this computed
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_as.cpp:1142]: why are we hardcoding these?

File: lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp (hunk 3 above)
Comment (VinayKumar1591) [reply lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp]: This was added to debug

File: lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp (hunk 4 above)
Comment (VinayKumar1591) [reply lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp]: this code not required, removed

File: lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp (hunk 5 above)
Comment (VinayKumar1591) [reply lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp]: reverted

**Review by ulaskozat — APPROVED** [review]
LGTM. TBD: IA and EA selection. They are still hardcoded here.

**Pull request merged**